# Day 1

In [1]:
import pandas as pd

print('Setup Berhasil...')
print(f'Versi pandas yang digunakan : {pd.__version__}')

Setup Berhasil...
Versi pandas yang digunakan : 2.2.2


In [2]:
# url data titanic
url = 'https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv'

# baca data dari url di atas
df = pd.read_csv(url)

# Tampilkan data
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

# Day 2 (Interogasi Data)

## Inspeksi Dasar

In [3]:
# lihat bentuk data
df.shape

(891, 12)

artinya ada 891 baris (penumpang) dan 12 kolom (informasi).

In [4]:
# Lihat ringkasan teknis dari data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Perhatikan kolom Age, Cabin, dan Embarked. Lihat di kolom Non-Null Count. Jumlahnya kurang dari 891. Ini petunjuk pertama: ada data yang hilang!

In [5]:
# lihat ringkasan statistik untuk kolom angka
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

rata-rata usia (mean di baris Age) sekitar 29.7 tahun, dan usia penumpang tertua adalah 80 tahun (max)

## Memilih data (Slicing dan Dicing)

In [6]:
# Memilih 1 kolom
df['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
886                                Montvila, Rev. Juozas
887                         Graham, Miss. Margaret Edith
888             Johnston, Miss. Catherine Helen "Carrie"
889                                Behr, Mr. Karl Howell
890                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [7]:
# Memilih beberapa kolom
df[['Name', 'Sex', 'Age']]

Name     Sex   Age
0                              Braund, Mr. Owen Harris    male  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0
2                               Heikkinen, Miss. Laina  female  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0
4                             Allen, Mr. William Henry    male  35.0
..                                                 ...     ...   ...
886                              Montvila, Rev. Juozas    male  27.0
887                       Graham, Miss. Margaret Edith  female  19.0
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN
889                              Behr, Mr. Karl Howell    male  26.0
890                                Dooley, Mr. Patrick    male  32.0

[891 rows x 3 columns]

In [8]:
# Memfilter baris
df[df['Age'] > 60]

PassengerId  Survived  Pclass                                       Name  \
33            34         0       2                      Wheadon, Mr. Edward H   
54            55         0       1             Ostby, Mr. Engelhart Cornelius   
96            97         0       1                  Goldschmidt, Mr. George B   
116          117         0       3                       Connors, Mr. Patrick   
170          171         0       1                  Van der hoef, Mr. Wyckoff   
252          253         0       1                  Stead, Mr. William Thomas   
275          276         1       1          Andrews, Miss. Kornelia Theodosia   
280          281         0       3                           Duane, Mr. Frank   
326          327         0       3                  Nysveen, Mr. Johan Hansen   
438          439         0       1                          Fortune, Mr. Mark   
456          457         0       1                  Millet, Mr. Francis Davis   
483          484         1       3                     Turkula, Mrs. (Hedwig)   
493          494         0       1                    Artagaveytia, Mr. Ramon   
545          546         0       1               Nicholson, Mr. Arthur Ernest   
555          556         0       1                         Wright, Mr. George   
570          571         1       2                         Harris, Mr. George   
625          626         0       1                      Sutton, Mr. Frederick   
630          631         1       1       Barkworth, Mr. Algernon Henry Wilson   
672          673         0       2                Mitchell, Mr. Henry Michael   
745          746         0       1               Crosby, Capt. Edward Gifford   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   
851          852         0       3                        Svensson, Mr. Johan   

        Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked  
33     male  66.0      0      0   C.A. 24579   10.5000          NaN        S  
54     male  65.0      0      1       113509   61.9792          B30        C  
96     male  71.0      0      0     PC 17754   34.6542           A5        C  
116    male  70.5      0      0       370369    7.7500          NaN        Q  
170    male  61.0      0      0       111240   33.5000          B19        S  
252    male  62.0      0      0       113514   26.5500          C87        S  
275  female  63.0      1      0        13502   77.9583           D7        S  
280    male  65.0      0      0       336439    7.7500          NaN        Q  
326    male  61.0      0      0       345364    6.2375          NaN        S  
438    male  64.0      1      4        19950  263.0000  C23 C25 C27        S  
456    male  65.0      0      0        13509   26.5500          E38        S  
483  female  63.0      0      0         4134    9.5875          NaN        S  
493    male  71.0      0      0     PC 17609   49.5042          NaN        C  
545    male  64.0      0      0          693   26.0000          NaN        S  
555    male  62.0      0      0       113807   26.5500          NaN        S  
570    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S  
625    male  61.0      0      0        36963   32.3208          D50        S  
630    male  80.0      0      0        27042   30.0000          A23        S  
672    male  70.0      0      0   C.A. 24580   10.5000          NaN        S  
745    male  70.0      1      1    WE/P 5735   71.0000          B22        S  
829  female  62.0      0      0       113572   80.0000          B28      NaN  
851    male  74.0      0      0       347060    7.7750          NaN        S

In [9]:
# Menggabungkan beberapa kondisi
df[(df['Sex'] == 'male') & (df['Age'] > 40)]

PassengerId  Survived  Pclass                            Name   Sex  \
6              7         0       1         McCarthy, Mr. Timothy J  male   
33            34         0       2           Wheadon, Mr. Edward H  male   
35            36         0       1  Holverson, Mr. Alexander Oskar  male   
54            55         0       1  Ostby, Mr. Engelhart Cornelius  male   
62            63         0       1     Harris, Mr. Henry Birkhardt  male   
..           ...       ...     ...                             ...   ...   
845          846         0       3             Abbing, Mr. Anthony  male   
851          852         0       3             Svensson, Mr. Johan  male   
857          858         1       1          Daly, Mr. Peter Denis   male   
860          861         0       3         Hansen, Mr. Claus Peter  male   
873          874         0       3     Vander Cruyssen, Mr. Victor  male   

      Age  SibSp  Parch      Ticket     Fare Cabin Embarked  
6    54.0      0      0       17463  51.8625   E46        S  
33   66.0      0      0  C.A. 24579  10.5000   NaN        S  
35   42.0      1      0      113789  52.0000   NaN        S  
54   65.0      0      1      113509  61.9792   B30        C  
62   45.0      1      0       36973  83.4750   C83        S  
..    ...    ...    ...         ...      ...   ...      ...  
845  42.0      0      0   C.A. 5547   7.5500   NaN        S  
851  74.0      0      0      347060   7.7750   NaN        S  
857  51.0      0      0      113055  26.5500   E17        S  
860  41.0      2      0      350026  14.1083   NaN        S  
873  47.0      0      0      345765   9.0000   NaN        S  

[102 rows x 12 columns]

## The First Insight

In [10]:
# Berapa jumlah penumpang berdasarkan jenis kelamin
df['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

.value_counts() adalah fungsi terbaik untuk menghitung data kategorika

In [11]:
# Berapa rata-rata tarif tiket
df['Fare'].mean()

32.204207968574636

In [12]:
# Siapa saja penumpang yang berada di kelas 1 dan selamat
df[(df['Pclass'] == 1) & (df['Survived'] == 1)].Name

1      Cumings, Mrs. John Bradley (Florence Briggs Th...
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
11                              Bonnell, Miss. Elizabeth
23                          Sloper, Mr. William Thompson
31        Spencer, Mrs. William Augustus (Marie Eugenie)
                             ...                        
862    Swift, Mrs. Frederick Joel (Margaret Welles Ba...
871     Beckwith, Mrs. Richard Leonard (Sallie Monypeny)
879        Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)
887                         Graham, Miss. Margaret Edith
889                                Behr, Mr. Karl Howell
Name: Name, Length: 136, dtype: object

# Day 3 (Data Cleaning & Feature Engineering)

## Persiapan & Identifikasi Masalah

In [14]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

kolom Age kehilangan banyak data, Cabin kehilangan sangat banyak data, dan Embarked kehilangan sedikit data

## Missing Value Imputation

### Strategi untuk Age (Kolom Numerik)

In [16]:
# Hitung usia rata-rata penumpang
mean_age = df['Age'].mean()
print(f'Rata-rata usia penumpang: {mean_age:.2f} tahun')

# Isi missing value Age dengan rata-rata penumpang
df['Age'].fillna(mean_age, inplace=True)

Rata-rata usia penumpang: 29.70 tahun


C:\Users\mulin\AppData\Local\Temp\ipykernel_28684\1042923696.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Age'].fillna(mean_age, inplace=True)


In [17]:
# Cek kembali apakah masih ada data yang hilang di kolom age
print(df['Age'].isnull().sum())

0


### Strategi untuk Embarked (Kolom Kategorikal)

In [19]:
# Cari modus dari Embarked
mode_embarked = df['Embarked'].mode()[0]

# Isi nilai NaN di kolom Emabarked dengan modusnya
df['Embarked'].fillna(mode_embarked, inplace=True)

# Cek kembali apakah masih ada data yang hilang di kolom Embarked
print(df['Embarked'].isnull().sum())

0


### Strategi untuk Cabin

Ide: Kolom ini kehilangan terlalu banyak data (lebih dari 75%). Mengisinya dengan nilai apapun kemungkinan besar akan menambah noise (gangguan) daripada sinyal. Untuk saat ini, keputusan terbaik adalah menghapusnya.

In [20]:
# Hapus kolom Cabin dari dataframe
df.drop('Cabin', axis=1, inplace=True)

# Lihat kolom yang tersisa
print(df.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Embarked'],
      dtype='object')


## Basic Feature Engineering

### Ide 1: Ukuran Keluarga.

In [23]:
# FamilySize = Jumlah saudara/pasangan (SibSp) + Jumlah orang tua/anak (Parch) + 1 (diri sendiri)
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

# Lihat hasilnya
df[['Name', 'SibSp', 'Parch', 'FamilySize']].head()

Name  SibSp  Parch  FamilySize
0                            Braund, Mr. Owen Harris      1      0           2
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      1      0           2
2                             Heikkinen, Miss. Laina      0      0           1
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      1      0           2
4                           Allen, Mr. William Henry      0      0           1

### Ide 2: Apakah bepergian sendiri?

In [26]:
# Buat kolom isAlone
import numpy as np
df['IsAlone'] = np.where(df['FamilySize'] == 1, 1, 0)

# Tampilkan
df[['Name', 'FamilySize', 'IsAlone']].head(10)

Name  FamilySize  IsAlone
0                            Braund, Mr. Owen Harris           2        0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...           2        0
2                             Heikkinen, Miss. Laina           1        1
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)           2        0
4                           Allen, Mr. William Henry           1        1
5                                   Moran, Mr. James           1        1
6                            McCarthy, Mr. Timothy J           1        1
7                     Palsson, Master. Gosta Leonard           5        0
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)           3        0
9                Nasser, Mrs. Nicholas (Adele Achem)           2        0

## Menangani Data Teks (Categorical Encoding)

Mengubah kolom teks (seperti 'Sex' dan 'Embarked') menjadi angka.

### Ubah kolom Sex

In [27]:
# Ubah male menjadi 0 dan female menjadi 1
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1      0   
2                             Heikkinen, Miss. Laina    1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1      0   
4                           Allen, Mr. William Henry    0  35.0      0      0   

             Ticket     Fare Embarked  FamilySize  isAlone  IsAlone  
0         A/5 21171   7.2500        S           2        0        0  
1          PC 17599  71.2833        C           2        0        0  
2  STON/O2. 3101282   7.9250        S           1        1        1  
3            113803  53.1000        S           2        0        0  
4            373450   8.0500        S           1        1        1

### Ubah kolom Embarked

In [28]:
embarked_dummies = pd.get_dummies(df['Embarked'], prefix='Embarked')

df = pd.concat([df, embarked_dummies], axis=1)

df.drop('Embarked', axis=1, inplace=True)

df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1      0   
2                             Heikkinen, Miss. Laina    1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1      0   
4                           Allen, Mr. William Henry    0  35.0      0      0   

             Ticket     Fare  FamilySize  isAlone  IsAlone  Embarked_C  \
0         A/5 21171   7.2500           2        0        0       False   
1          PC 17599  71.2833           2        0        0        True   
2  STON/O2. 3101282   7.9250           1        1        1       False   
3            113803  53.1000           2        0        0       False   
4            373450   8.0500           1        1        1       False   

   Embarked_Q  Embarked_S  
0       False        True  
1       False       False  
2       False        True  
3       False        True  
4       False        True